src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java

```diff
@@ -0,0 +1,173 @@
+    private SequenceType _sequenceType;
+
+    private ItemType itemType;
+    private String sequenceTypeName;
+
+    private Item _nextResult;
+    private int _childIndex;
+
+    public TypePromotionIterator(
+            String exceptionMessage,
+            RuntimeIterator iterator,
+            SequenceType sequenceType,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(Collections.singletonList(iterator), iteratorMetadata);
+        this._exceptionMessage = exceptionMessage;
+        this._iterator = iterator;
+        this._sequenceType = sequenceType;
+        itemType = _sequenceType.getItemType();
+        sequenceTypeName = ItemTypes.getItemTypeName(itemType.getType().toString());
+    }
+
+    @Override
+    public boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    public void resetLocal(DynamicContext context) {
+        _iterator.reset(_currentDynamicContext);
+        setNextResult();
+    }
+
+    @Override
+    public void closeLocal() {
+        _iterator.close();
+    }
+
+    @Override
+    public void openLocal() {
+        this._childIndex = 0;
+        _iterator.open(_currentDynamicContext);
+        this.setNextResult();
+    }
+
+
+    @Override
+    public Item nextLocal() {
+        if (this._hasNext) {
+            Item _currentResult = _nextResult;
+            setNextResult();
+            return _currentResult;
+        } else
+            throw new IteratorFlowException(RuntimeIterator.FLOW_EXCEPTION_MESSAGE, getMetadata());
+    }
+
+    private void setNextResult() {
+        _nextResult = null;
+        if (_iterator.hasNext()) {
+            _nextResult = _iterator.next();
+            if (_nextResult != null)
+                _childIndex++;
+        } else {
+            _iterator.close();
+            checkEmptySequence(_childIndex);
+        }
+
+        this._hasNext = _nextResult != null;
+        if (!hasNext())
+            return;
+
+        checkItemsSize(_childIndex);
+        if (!_nextResult.isTypeOf(itemType)) {
+            checkTypePromotion();
+        }
+    }
+
+    private void checkEmptySequence(long size) {
+        if (
+            size == 0
+                && (_sequenceType.getArity() == SequenceType.Arity.One
+                    ||
+                    _sequenceType.getArity() == SequenceType.Arity.OneOrMore)
+        ) {
+            throw new UnexpectedTypeException(
+                    "Empty sequence cannot be promoted to type "
+                        + sequenceTypeName
+                        + _sequenceType.getArity().getSymbol(),
+                    getMetadata()
+            );
+        }
+    }
+
+    private void checkItemsSize(long size) {
+        if (size > 0 && _sequenceType.isEmptySequence()) {
+            throw new UnexpectedTypeException(
+                    ItemTypes.getItemTypeName(_nextResult.getClass().getSimpleName())
+                        + " cannot be promoted to type empty-sequence()",
```

Comment:
Same here: "Expecting empty sequence, but the value provided has at least one item."